File: DataStructureJava/tree/RBTree.java
import java.util.ArrayList;

public class RBTree <K extends Comparable<K>, V>{
    private static final boolean RED = true;
    private static final boolean BLACK = false;

    private class Node{
        public K key;
        public V value;
        public Node left,right;
        public boolean color;

        public Node(K key, V value){
            this.key = key;
            this.value = value;
            left = null;
            right = null;
            color = RED;
        }
    }
    private Node root;
    private int size;

    public RBTree(){
        root = null;
        size = 0;
    }
    public int size(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public boolean isRed(Node node){
        if(node == null)
            return BLACK;
        return node.color;
    }
    //对节点node进行向左旋转操作。返回旋转后的新的根节点x
    //  node                               x
//     /  \                               /  \
//    T1   x                            node  T3
//        /  \        左旋转>>>>        /  \
//       T2  T3                        T1  T2

    private Node leftRotate(Node node){
        Node x = node.right;

        // 左旋转
        node.right = x.left;
        x.left = node;

        x.color = node.color;
        node.color = RED;
        return x;
    }
    //对节点node进行向右旋转操作。返回旋转后的新的根节点x
    //  node                               x
//     /  \                               /  \
//    x   T2                             y   node
//   /  \        右旋转>>>>                   /  \
//  y   T1                                  T1  T2

    private Node rightRotate(Node node){
        Node x = node.left;

        // 右旋转
        node.left = x.right;
        x.right = node;

        x.color = node.color;
        node.color = RED;
        return x;
    }
    //颜色翻转
    private void filpColors(Node node){
        node.color = RED;
        node.left.color = BLACK;
        node.right.color = BLACK;
    }

    public void add(K key, V value){
        root = add(root,key,value);
        root.color = BLACK; //最终根节点为黑色
    }

    // 以node为根节点，返回插入元素(key, value)后新的红黑树树的根
    private Node add(Node node,K key,V value){
        if (node == null){
            size ++;
            return new Node(key, value);
        }

        if (key.compareTo(node.key) < 0)
            node.left = add(node.left,key, value);
        else if(key.compareTo(node.key) > 0)
            node.right = add(node.right,key,value);
        else
            node.value = value;

        if(isRed(node.right)  && !isRed(node.left))
            node = leftRotate(node);

        if(isRed(node.left)  && isRed(node.left.left))
            node = rightRotate(node);

        if(isRed(node.left)  && isRed(node.right))
            filpColors(node);

        return node;
    }
    @Override
    public String toString(){
        StringBuilder res = new StringBuilder();
        generateBSTString(root,0,res);
        return res.toString();
    }
    private void generateBSTString(Node node,int depth,StringBuilder res){
        if(node == null){
            res.append(generatDepthString(depth) + "null\n");
            return;
        }
        res.append(generatDepthString(depth) + node.key +"\n");
        generateBSTString(node.left,depth+1,res);
        generateBSTString(node.right,depth+1,res);

    }
    private String generatDepthString(int depth){
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < depth; i ++)
            res.append("--");
        return res.toString();
    }

}

File: data_structure_python/tree/RBTree.py
"""A red_black tree is a binary tree that satisfies the following red-black properties:
1, Every node is either red or black
2, The root is black
3, Every leaf (and node is none) is black
4, If a node is red, then both its children are black
5, For each node, all simple paths from the node to descendant leaves contain the same number of black nodes
红黑树是保持"黑平衡"的二叉树
"""

RED = True
BLACK = False


class TreeNode:
    def __init__(self, key, val, color=RED):
        self.key = key
        self.val = val
        self.color = color
        self.left = None
        self.right = None
    def __str__(self):
        return "{}:{}".format(self.key,self.val)

class RBTree:
    def __init__(self):
        self.root = None
        self.size = 0
    def __len__(self):
        return self.size
    def is_red(self, node):
        if node is None:
            return BLACK
        return node.color

    # 对节点node进行向左旋转操作。返回旋转后的新的根节点x
    #       node                       x
    #      /   \                      /  \
    #     T1    x                  node    z
    #          /  \   左旋转>>>>   /  \
    #         T2   z             T1  T2
    def _left_rotate(self, node):
        x = node.right

        # left rotate
        node.right = x.left
        x.left = node

        x.color = node.color
        node.color = RED
        return x

    # 对节点node进行向右旋转操作。返回旋转后的新的根节点x
    #       node                     x
    #      /   \                   /    \
    #     x     T4                z     node
    #    /  \        右旋转>>>>          /  \
    #   z   T3                         T3  T4

    def _right_rotate(self, node):
        x = node.left
        # right rotate
        node.left = x.right
        x.right = node

        x.color = node.color
        node.color = RED
        return x

    def _flip_colors(self, node):
        node.color = RED
        node.left.color = BLACK
        node.right.color = BLACK

    def add(self, key, val):
        self.root = self._add(self.root, key, val)
        self.root.color = BLACK

    def _add(self, node, key, val):
        if node is None:
            self.size += 1
            return TreeNode(key, val)

        if key > node.key:
            node.right = self._add(node.right, key, val)
        elif key < node.key:
            node.left = self._add(node.left, key, val)
        else:
            node.val = val

        # 维护黑平衡
        if self.is_red(node.right) and not self.is_red(node.left):
            node = self._left_rotate(node)

        if self.is_red(node.left) and self.is_red(node.left.left):
            node = self._right_rotate(node)

        if self.is_red(node.left) and self.is_red(node.right) :
            self._flip_colors(node)

        return node

    def _inorder(self, node, dic):
        if node is None:
            return
        self._inorder(node.left, dic)
        dic[node.key] = node.val
        self._inorder(node.right, dic)

    def __str__(self):
        rest = {}
        self._inorder(self.root, rest)
        return str(rest)


if __name__ == '__main__':
    nodes = [5, 1, 3, 9, 4]
    RBT = RBTree()

    for node in nodes:
        RBT.add(node, node)
    print(RBT)
    print(len(RBT))
